File: CS/Q283154/Bootstrapper.cs
using Microsoft.Practices.Prism.MefExtensions;
using System.ComponentModel.Composition.Hosting;
using System.Windows;
using Microsoft.Practices.Prism.Modularity;
using System.ComponentModel.Composition;
using Microsoft.Practices.Prism.Regions;
using DevExpress.Xpf.Ribbon;
namespace Q283154 {
    public class Bootstrapper : MefBootstrapper {
        protected override void ConfigureAggregateCatalog() {
            base.ConfigureAggregateCatalog();
            this.AggregateCatalog.Catalogs.Add(new AssemblyCatalog(typeof(Bootstrapper).Assembly));
        }
        protected override void ConfigureModuleCatalog() {
            base.ConfigureModuleCatalog();
            ModuleCatalog moduleCatalog = (ModuleCatalog)this.ModuleCatalog;
            moduleCatalog.AddModule(typeof(Module1));
        }
        protected override DependencyObject CreateShell() {
            return new Shell();
        }
        protected override void InitializeShell() {
            base.InitializeShell();
            Application.Current.RootVisual = (Shell)this.Shell;
        }
        protected override RegionAdapterMappings ConfigureRegionAdapterMappings() {
            RegionAdapterMappings mappings = base.ConfigureRegionAdapterMappings();
            mappings.RegisterMapping(typeof(RibbonPageCategoryBase), Container.GetExportedValue<RibbonAdapter>());
            return mappings;
        }
    }

    [Export]
    [PartCreationPolicy(CreationPolicy.NonShared)]
    public class RibbonAdapter : RegionAdapterBase<RibbonPageCategoryBase> {
        [ImportingConstructor]
        public RibbonAdapter(IRegionBehaviorFactory behaviorFactory)
            : base(behaviorFactory) {
        }
        protected override IRegion CreateRegion() {
            return new Region();
        }
        protected override void Adapt(IRegion region, RibbonPageCategoryBase regionTarget) {
            new Manager(region, regionTarget);
        }
        class Manager {
            IRegion Region;
            RibbonPageCategoryBase RegionTarget;
            public Manager(IRegion region, RibbonPageCategoryBase regionTarget) {
                Region = region;
                RegionTarget = regionTarget;
                Region.Views.CollectionChanged += new System.Collections.Specialized.NotifyCollectionChangedEventHandler(Views_CollectionChanged);
            }

            void Views_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) {
                if(e.Action == System.Collections.Specialized.NotifyCollectionChangedAction.Reset) {
                    RegionTarget.Pages.Clear();
                    foreach(RibbonPage p in Region.Views)
                        RegionTarget.Pages.Add(p);
                }
            }
        }
    }
         
}

File: VB/Q283154/Bootstrapper.vb
Imports Microsoft.VisualBasic
Imports Microsoft.Practices.Prism.MefExtensions
Imports System.ComponentModel.Composition.Hosting
Imports System.Windows
Imports Microsoft.Practices.Prism.Modularity
Imports System.ComponentModel.Composition
Imports Microsoft.Practices.Prism.Regions
Imports DevExpress.Xpf.Ribbon
Namespace Q283154
	Public Class Bootstrapper
		Inherits MefBootstrapper
		Protected Overrides Sub ConfigureAggregateCatalog()
			MyBase.ConfigureAggregateCatalog()
			Me.AggregateCatalog.Catalogs.Add(New AssemblyCatalog(GetType(Bootstrapper).Assembly))
		End Sub
		Protected Overrides Sub ConfigureModuleCatalog()
			MyBase.ConfigureModuleCatalog()
			Dim moduleCatalog As ModuleCatalog = CType(Me.ModuleCatalog, ModuleCatalog)
			moduleCatalog.AddModule(GetType(Module1))
		End Sub
		Protected Overrides Function CreateShell() As DependencyObject
			Return New Shell()
		End Function
		Protected Overrides Sub InitializeShell()
			MyBase.InitializeShell()
			Application.Current.RootVisual = CType(Me.Shell, Shell)
		End Sub
		Protected Overrides Function ConfigureRegionAdapterMappings() As RegionAdapterMappings
			Dim mappings As RegionAdapterMappings = MyBase.ConfigureRegionAdapterMappings()
			mappings.RegisterMapping(GetType(RibbonPageCategoryBase), Container.GetExportedValue(Of RibbonAdapter)())
			Return mappings
		End Function
	End Class

	<Export, PartCreationPolicy(CreationPolicy.NonShared)> _
	Public Class RibbonAdapter
		Inherits RegionAdapterBase(Of RibbonPageCategoryBase)
		<ImportingConstructor> _
		Public Sub New(ByVal behaviorFactory As IRegionBehaviorFactory)
			MyBase.New(behaviorFactory)
		End Sub
		Protected Overrides Function CreateRegion() As IRegion
			Return New Region()
		End Function
		Protected Overrides Sub Adapt(ByVal region As IRegion, ByVal regionTarget As RibbonPageCategoryBase)
			Dim TempManager As Manager = New Manager(region, regionTarget)
		End Sub
		Private Class Manager
			Private Region As IRegion
			Private RegionTarget As RibbonPageCategoryBase
			Public Sub New(ByVal region As IRegion, ByVal regionTarget As RibbonPageCategoryBase)
				Me.Region = region
				Me.RegionTarget = regionTarget
				AddHandler Me.Region.Views.CollectionChanged, AddressOf Views_CollectionChanged
			End Sub

			Private Sub Views_CollectionChanged(ByVal sender As Object, ByVal e As System.Collections.Specialized.NotifyCollectionChangedEventArgs)
				If e.Action = System.Collections.Specialized.NotifyCollectionChangedAction.Reset Then
					RegionTarget.Pages.Clear()
					For Each p As RibbonPage In Region.Views
						RegionTarget.Pages.Add(p)
					Next p
				End If
			End Sub
		End Class
	End Class

End Namespace